# Project Title
### Data Engineering Capstone Project

#### Project Summary
--describe your project at a high level--

The project follows the follow steps:
* Step 1: Scope the Project and Gather Data
* Step 2: Explore and Assess the Data
* Step 3: Define the Data Model
* Step 4: Run ETL to Model the Data
* Step 5: Complete Project Write Up

In [1]:
# Do all imports and installs here
import pandas as pd

### Step 1: Scope the Project and Gather Data

#### Scope 
Explain what you plan to do in the project in more detail. What data do you use? What is your end solution look like? What tools did you use? etc>

#### Describe and Gather Data 
Describe the data sets you're using. Where did it come from? What type of information is included? 

In [2]:
# Read in the data here
taxi_file_url = "https://data.cityofnewyork.us/api/views/pkmi-4kfn/rows.csv"
taxi_zone_file = "taxi_zones.json"
weather_file = "weather.csv"

df_zones = pd.read_json(taxi_zone_file)
df_weather = pd.read_csv(weather_file)
df_taxi = pd.read_csv(taxi_file_url)

/Users/arnavgohil/.conda/envs/Capstone-Project/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
df_taxi.head()

VendorID    lpep_pickup_datetime   lpep_dropoff_datetime  \
0       2.0  12/18/2019 03:52:30 PM  12/18/2019 03:54:39 PM   
1       2.0  01/01/2020 12:45:58 AM  01/01/2020 12:56:39 AM   
2       2.0  01/01/2020 12:41:38 AM  01/01/2020 12:52:49 AM   
3       1.0  01/01/2020 12:52:46 AM  01/01/2020 01:14:21 AM   
4       1.0  01/01/2020 12:19:57 AM  01/01/2020 12:30:56 AM   

  store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  passenger_count  \
0                  N         1.0           264           264              5.0   
1                  N         5.0            66            65              2.0   
2                  N         1.0           181           228              1.0   
3                  N         1.0           129           263              2.0   
4                  N         1.0           210           150              1.0   

   trip_distance  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           0.00          3.5   0.50      0.5        0.01           0.0   
1           1.28         20.0   0.00      0.0        4.06           0.0   
2           2.47         10.5   0.50      0.5        3.54           0.0   
3           6.30         21.0   3.25      0.5        0.00           0.0   
4           2.30         10.0   0.50      0.5        0.00           0.0   

   ehail_fee  improvement_surcharge  total_amount  payment_type  trip_type  \
0        NaN                    0.3          4.81           1.0        1.0   
1        NaN                    0.3         24.36           1.0        2.0   
2        NaN                    0.3         15.34           1.0        1.0   
3        NaN                    0.3         25.05           2.0        1.0   
4        NaN                    0.3         11.30           1.0        1.0   

   congestion_surcharge  
0                  0.00  
1                  0.00  
2                  0.00  
3                  2.75  
4                  0.00

In [4]:
df_zones.head()

LocationID        Borough                     Zone service_zone
0           1            EWR           Newark Airport          EWR
1           2         Queens              Jamaica Bay    Boro Zone
2           3          Bronx  Allerton/Pelham Gardens    Boro Zone
3           4      Manhattan            Alphabet City  Yellow Zone
4           5  Staten Island            Arden Heights    Boro Zone

In [5]:
df_weather.head()

STATION                       NAME  LATITUDE  LONGITUDE  ELEVATION  \
0  USW00014732  LA GUARDIA AIRPORT, NY US   40.7792     -73.88        3.4   
1  USW00014732  LA GUARDIA AIRPORT, NY US   40.7792     -73.88        3.4   
2  USW00014732  LA GUARDIA AIRPORT, NY US   40.7792     -73.88        3.4   
3  USW00014732  LA GUARDIA AIRPORT, NY US   40.7792     -73.88        3.4   
4  USW00014732  LA GUARDIA AIRPORT, NY US   40.7792     -73.88        3.4   

             DATE  HLY-CLDH-NORMAL  HLY-CLOD-PCTBKN  HLY-CLOD-PCTCLR  \
0  01-01T01:00:00              0.0              134              268   
1  01-01T02:00:00              0.0              155              277   
2  01-01T03:00:00              0.0              162              277   
3  01-01T04:00:00              0.0              149              262   
4  01-01T05:00:00              0.0              135              243   

   HLY-CLOD-PCTFEW  ...  HLY-TEMP-NORMAL  HLY-WCHL-NORMAL  HLY-WIND-1STDIR  \
0               75  ...              1.4             -3.2                8   
1               79  ...              1.2             -3.4                8   
2               63  ...              1.1             -3.6                8   
3               68  ...              0.9             -3.8                8   
4              119  ...              0.8             -3.9                8   

   HLY-WIND-1STPCT  HLY-WIND-2NDDIR  HLY-WIND-2NDPCT  HLY-WIND-AVGSPD  \
0              269                6              178              5.6   
1              251                7              197              5.6   
2              282                7              182              5.5   
3              284                7              164              5.5   
4              273                7              174              5.5   

   HLY-WIND-PCTCLM  HLY-WIND-VCTDIR  HLY-WIND-VCTSPD  
0               24              308              2.8  
1               33              309              2.9  
2               27              312              2.9  
3               29              314              2.9  
4               31              314              2.9  

[5 rows x 32 columns]

In [6]:
rate_data = { 'RatecodeID':[1,2,3,4,5,6],
              'Rate_type': ['Standard rate','JFK','Newark','Nassau or Westchester','Negotiated fare','Group ride']}
df_rate = pd.DataFrame(rate_data)
df_rate.head()

RatecodeID              Rate_type
0           1          Standard rate
1           2                    JFK
2           3                 Newark
3           4  Nassau or Westchester
4           5        Negotiated fare

In [7]:
payment_data = { 'PaymentID':[1,2,3,4,5,6],
                 'Payment_type': ['Credit card','Cash','No charge','Dispute','Unknown','Voided trip']}
df_payment = pd.DataFrame(payment_data)
df_payment.head()

PaymentID Payment_type
0          1  Credit card
1          2         Cash
2          3    No charge
3          4      Dispute
4          5      Unknown

In [8]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import udf, col
from pyspark.sql.functions import year, month, dayofmonth, hour, weekofyear, date_format

In [9]:
spark = SparkSession \
        .builder \
        .config("spark.jars.packages", "org.apache.hadoop:hadoop-aws:2.7.0") \
        .getOrCreate()

In [10]:
spark

In [11]:
weather_spark_df = spark.read.option("multiline","true").option("header", "true").csv(weather_file)
weather_spark_df.head(1)

[Row(STATION='USW00014732', NAME='LA GUARDIA AIRPORT, NY US', LATITUDE='40.7792', LONGITUDE='-73.88', ELEVATION='3.4', DATE='01-01T01:00:00', HLY-CLDH-NORMAL='0.0', HLY-CLOD-PCTBKN='134', HLY-CLOD-PCTCLR='268', HLY-CLOD-PCTFEW='75', HLY-CLOD-PCTOVC='438', HLY-CLOD-PCTSCT='86', HLY-DEWP-10PCTL='-14.4', HLY-DEWP-90PCTL='5.0', HLY-DEWP-NORMAL='-4.7', HLY-HIDX-NORMAL='1.4', HLY-HTDH-NORMAL='17.0', HLY-PRES-10PCTL='1005.3', HLY-PRES-90PCTL='1031.0', HLY-PRES-NORMAL='1018.2', HLY-TEMP-10PCTL='-5.6', HLY-TEMP-90PCTL='7.8', HLY-TEMP-NORMAL='1.4', HLY-WCHL-NORMAL='-3.2', HLY-WIND-1STDIR='8', HLY-WIND-1STPCT='269', HLY-WIND-2NDDIR='6', HLY-WIND-2NDPCT='178', HLY-WIND-AVGSPD='5.6', HLY-WIND-PCTCLM='24', HLY-WIND-VCTDIR='308', HLY-WIND-VCTSPD='2.8')]

In [12]:
zone_spark_df = spark.read.option("multiline","true").json(taxi_zone_file)
zone_spark_df.head(1)

[Row(Borough='EWR', LocationID=1, Zone='Newark Airport', service_zone='EWR')]

In [13]:
df_taxi.head(2)

VendorID    lpep_pickup_datetime   lpep_dropoff_datetime  \
0       2.0  12/18/2019 03:52:30 PM  12/18/2019 03:54:39 PM   
1       2.0  01/01/2020 12:45:58 AM  01/01/2020 12:56:39 AM   

  store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  passenger_count  \
0                  N         1.0           264           264              5.0   
1                  N         5.0            66            65              2.0   

   trip_distance  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           0.00          3.5    0.5      0.5        0.01           0.0   
1           1.28         20.0    0.0      0.0        4.06           0.0   

   ehail_fee  improvement_surcharge  total_amount  payment_type  trip_type  \
0        NaN                    0.3          4.81           1.0        1.0   
1        NaN                    0.3         24.36           1.0        2.0   

   congestion_surcharge  
0                   0.0  
1                   0.0

In [14]:
from pyspark.sql.types import *

taxiSchema = StructType([StructField("VendorID", FloatType(), True)\
                       ,StructField("lpep_pickup_datetime", StringType(), True)\
                       ,StructField("lpep_dropoff_datetime", StringType(), True)\
                       ,StructField("store_and_fwd_flag", StringType(), True)\
                       ,StructField("RatecodeID", FloatType(), True)\
                       ,StructField("PULocationID", IntegerType(), True)\
                       ,StructField("DOLocationID", IntegerType(), True)\
                       ,StructField("passenger_count", FloatType(), True)\
                       ,StructField("trip_distance", FloatType(), True)\
                       ,StructField("fare_amount", FloatType(), True)\
                       ,StructField("extra", FloatType(), True)\
                       ,StructField("mta_tax", FloatType(), True)\
                       ,StructField("tip_amount", FloatType(), True)\
                       ,StructField("tolls_amount", FloatType(), True)\
                       ,StructField("ehail_fee", FloatType(), True)\
                       ,StructField("improvement_surcharge", FloatType(), True)\
                       ,StructField("total_amount", FloatType(), True)\
                       ,StructField("payment_type", FloatType(), True)\
                       ,StructField("trip_type", FloatType(), True)\
                       ,StructField("congestion_surcharge", FloatType(), True)])


In [16]:
taxi_spark_df = spark.createDataFrame(df_taxi,schema=taxiSchema)
taxi_spark_df.head(2)

[Row(VendorID=2.0, lpep_pickup_datetime='12/18/2019 03:52:30 PM', lpep_dropoff_datetime='12/18/2019 03:54:39 PM', store_and_fwd_flag='N', RatecodeID=1.0, PULocationID=264, DOLocationID=264, passenger_count=5.0, trip_distance=0.0, fare_amount=3.5, extra=0.5, mta_tax=0.5, tip_amount=0.009999999776482582, tolls_amount=0.0, ehail_fee=nan, improvement_surcharge=0.30000001192092896, total_amount=4.809999942779541, payment_type=1.0, trip_type=1.0, congestion_surcharge=0.0),
 Row(VendorID=2.0, lpep_pickup_datetime='01/01/2020 12:45:58 AM', lpep_dropoff_datetime='01/01/2020 12:56:39 AM', store_and_fwd_flag='N', RatecodeID=5.0, PULocationID=66, DOLocationID=65, passenger_count=2.0, trip_distance=1.2799999713897705, fare_amount=20.0, extra=0.0, mta_tax=0.0, tip_amount=4.059999942779541, tolls_amount=0.0, ehail_fee=nan, improvement_surcharge=0.30000001192092896, total_amount=24.360000610351562, payment_type=1.0, trip_type=2.0, congestion_surcharge=0.0)]

In [19]:
# Rate Spark DF
rateSchema = StructType([StructField("RatecodeID",IntegerType(),True),StructField("Rate_type",StringType(),True)])
rate_spark_df = spark.createDataFrame(df_rate,schema=rateSchema)
rate_spark_df.head(2)

[Row(RatecodeID=1, Rate_type='Standard rate'),
 Row(RatecodeID=2, Rate_type='JFK')]

In [20]:
# Payment Spark DF
paymentSchema = StructType([StructField("PaymentID",IntegerType(),True),StructField("Payment_type",StringType(),True)])
payment_spark_df = spark.createDataFrame(df_payment,schema=paymentSchema)
payment_spark_df.head(2)

[Row(PaymentID=1, Payment_type='Credit card'),
 Row(PaymentID=2, Payment_type='Cash')]

### Step 2: Explore and Assess the Data
#### Explore the Data 
Identify data quality issues, like missing values, duplicate data, etc.

1. Remove rows where Payment ID is Unknown
2. Remove rows where PickUpID or DropID is 264 or 265, check all values in range.
3. Remove rows which are out of range in taxi
4. Remove rows with Missing Columns in taxi data or is NULL/NaN
5. Filter Weather data so that it matches time period of taxi data
6. Remove duplicate data in weather. (keep only JFK, remove others)

#### Cleaning Steps
Document steps necessary to clean the data

In [ ]:
# Performing cleaning tasks here



### Step 3: Define the Data Model
#### 3.1 Conceptual Data Model
Map out the conceptual data model and explain why you chose that model

#### 3.2 Mapping Out Data Pipelines
List the steps necessary to pipeline the data into the chosen data model

### Step 4: Run Pipelines to Model the Data 
#### 4.1 Create the data model
Build the data pipelines to create the data model.

In [ ]:
# Write code here

#### 4.2 Data Quality Checks
Explain the data quality checks you'll perform to ensure the pipeline ran as expected. These could include:
 * Integrity constraints on the relational database (e.g., unique key, data type, etc.)
 * Unit tests for the scripts to ensure they are doing the right thing
 * Source/Count checks to ensure completeness
 
Run Quality Checks

In [ ]:
# Perform quality checks here

#### 4.3 Data dictionary 
Create a data dictionary for your data model. For each field, provide a brief description of what the data is and where it came from. You can include the data dictionary in the notebook or in a separate file.

#### Step 5: Complete Project Write Up
* Clearly state the rationale for the choice of tools and technologies for the project.
* Propose how often the data should be updated and why.
* Write a description of how you would approach the problem differently under the following scenarios:
 * The data was increased by 100x.
 * The data populates a dashboard that must be updated on a daily basis by 7am every day.
 * The database needed to be accessed by 100+ people.